nilearn/decoding/decoder.py

```diff
@@ -0,0 +1,857 @@
+            Note that this parameter must be specified in some scikit-learn
+            cross-validation generators to calculate the number of splits, e.g.
+            sklearn.model_selection.LeaveOneGroupOut or
+            sklearn.model_selection.LeavePGroupsOut.
+
+            For more details see https://scikit-learn.org/stable/modules/cross_validation.html#cross-validation-iterators-for-grouped-data
+
+        Attributes
+        ----------
+        `masker_`: instance of NiftiMasker or MultiNiftiMasker
+            The NiftiMasker used to mask the data.
+
+        `mask_img_`: Nifti1Image
+            Mask computed by the masker object.
+
+        `classes_`: numpy.ndarray
+            Classes to predict. For classification only.
+
+        `screening_percentile_`: float
+            Screening percentile corrected according to volume of mask,
+            relative to the volume of standard brain.
+
+        `coef_`: numpy.ndarray, shape=(n_classes, n_features)
+            Contains the mean of the models weight vector across
+            fold for each class.
+
+        `coef_img_`: dict of Nifti1Image
+            Dictionary containing `coef_` with class names as keys,
+            and `coef_` transformed in Nifti1Images as values. In the case
+            of a regression, it contains a single Nifti1Image at the key 'beta'.
+
+        `intercept_`: narray, shape (nclasses,)
+            Intercept (a.k.a. bias) added to the decision function.
+
+        `cv_`: list of pairs of lists
+            List of the (n_folds,) folds. For the corresponding fold,
+            each pair is composed of two lists of indices,
+            one for the train samples and one for the test samples.
+
+        `std_coef_`: numpy.ndarray, shape=(n_classes, n_features)
+            Contains the standard deviation of the models weight vector across
+            fold for each class. Note that folds are not independent, see
+            https://scikit-learn.org/stable/modules/cross_validation.html#cross-validation-iterators-for-grouped-data
+
+        `std_coef_img_`: dict of Nifti1Image
+            Dictionary containing `std_coef_` with class names as keys,
+            and `coef_` transformed in Nifti1Image as values. In the case
+            of a regression, it contains a single Nifti1Image at the key 'beta'.
+
+        `cv_params_`: dict of lists
+            Best point in the parameter grid for each tested fold
+            in the inner cross validation loop.
+
+        `cv_scores_`: dict, (classes, n_folds)
+            Scores (misclassification) for each parameter, and on each fold
+        """
+        self.estimator = _check_estimator(self.estimator)
+        self.memory_ = _check_memory(self.memory, self.verbose)
+
+        X = self._apply_mask(X)
+        X, y = check_X_y(X, y, dtype=np.float, multi_output=True)
+
+        # Setup scorer
+        scorer = check_scoring(self.estimator, self.scoring)
+        self.cv_ = list(check_cv(self.cv, y=y,
```

Comment:
if cv is an int or None check_cv will return a `StratifiedKFold`, which ignores the groups. https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.StratifiedKFold.html#sklearn.model_selection.StratifiedKFold.split

so I don't think @bthirion 's comment on spreading subjects over train and test has not been addressed?

Reply:
Yes, I agree with you. However, in that case is it better to point to the user to create a custom cross-validator object in scikit-learn, e.g. in the `plot_haxby_dfifferent_estimators.py`? 
https://github.com/nilearn/nilearn/blob/4f194b424efc641bcb1b73f5ff28d4ab7e1f7ee5/examples/02_decoding/plot_haxby_different_estimators.py#L68-L98

Reply:
At the very least, we need a detailed warning here.

I am thinking if `cv` is an `int` or `None`, and `groups` is specified, this should raise an error with a good explanation. This should also be mentioned in the docstring.

Reply:
I think that API-wise we should stick to scikit-learn, but we should insist in the doc on the choice of well-suited cross-validation: LeavePGroupsOut is generally the right choice.
The idea is that the user should feel responsible about the cv being right.  There are many possible cases, it is hard to plan for all of them in advance.

Reply:
I agree, but at the moment if the user provides a `groups` parameter and leaves the default cv, the groups are silently ignored. As I discussed IRL with @tbng yesterday a more appropriate cv generator from scikit-learn should be used when the user provides groups

Reply:
+1 with avoiding silent ignoring the groups. People stumble on that.